### PR TITLE
feat: decouple authentication logic from `useMetametrics`

### PR DESCRIPTION
**Pull request by mathieuartu**
## **Description**

This PR decouples the authentication logic that was incorporated into the `useMetametrics` hook.
It now lives in `metamask-controller.js`.

Auth sign-in / sign-out will now be performed whenever the `MetaMetricsController`'s state changes to signify that a user changed the value of `participateInMetaMetrics`

Scenarios: 

- Sign-in will be performed if a user enables either metametrics or profile syncing
- Sign-out will be performed if a user disables metametrics & has profile syncing turned off
- Alternatively, sign-out will be performed if a user disables profile syncing & has metametrics turned off

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/29070?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to the settings page
2. Turn off profile syncing and metametrics
3. Turn only metametrics back on
4. Look into the network tab and see requests to `https://authentication.api.cx.metamask.io`

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by mirceanis — APPROVED** [review]

File: app/scripts/metamask-controller.test.js [hunk 1]
```diff
@@ -2840,4 +2840,144 @@ describe('MetaMaskController', () => {
       expect(browserPolyfillMock.storage.session.set).not.toHaveBeenCalled();
     });
   });
+
+  describe('MetaMetrics & authentication dependencies', () => {
+    let metamaskController;
+    let mockPerformSignIn;
+    let mockPerformSignOut;
```
Comment (Prithpal-Sooriya) [inline app/scripts/metamask-controller.test.js:2847]: Nit (cleanup) - We don't need mutating variables, instead we can just return them in our arrange function.

```ts
const arrangeControllerState = (stateOverride) => {
  // ...
  return {
    metamaskController,
    mockPerformSignIn,
    mockPerformSignOut,
  }
}

```

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -1678,6 +1678,39 @@ export default class MetamaskController extends EventEmitter {
       }),
     });
 
+    this.controllerMessenger.subscribe(
+      'MetaMetricsController:stateChange',
+      previousValueComparator(async (prevState, currState) => {
```
[the remark this answers is not in the document]
Comment (Prithpal-Sooriya) [reply app/scripts/metamask-controller.js:1683]: This is a nit, and can be resolved.

File: app/scripts/metamask-controller.test.js [hunk 3]
```diff
@@ -2840,4 +2840,144 @@ describe('MetaMaskController', () => {
...
+          participateInMetaMetrics: false,
+        },
+      );
+
+      expect(mockPerformSignIn).not.toHaveBeenCalled();
+      expect(mockPerformSignOut).toHaveBeenCalledTimes(1);
+    });
+
+    it('should not sign in the user if MetaMetrics is enabled and the user is already signed in', async () => {
+      arrangeControllerState({
```
Comment (Prithpal-Sooriya) [inline app/scripts/metamask-controller.test.js:2918]: I really appreciate these tests! I was wondering if we could clean em up using a test matrix + `it.each`, but I think the setup required to run the test would be a bit messy. These are fine as is.

File: ui/hooks/identity/useProfileSyncing/profileSyncing.ts [hunk 4]
```diff
@@ -63,6 +68,11 @@ export function useDisableProfileSyncing(): {
       // disable profile syncing
       await dispatch(disableProfileSyncingAction());
 
+      // sign out the user if MetaMetrics is not enabled and the user is signed in
+      if (!isMetaMetricsEnabled && isSignedIn) {
+        await dispatch(performSignOut());
+      }
```
Comment (Prithpal-Sooriya) [inline ui/hooks/identity/useProfileSyncing/profileSyncing.ts:74]: Thanks for covering this!

File: ui/hooks/useMetametrics.ts [hunk 5]
```diff
@@ -35,10 +29,6 @@ export function useEnableMetametrics(): {
     setError(null);
 
     try {
-      if (isProfileSyncingEnabled && !isUserSignedIn) {
-        await dispatch(performSignIn());
-      }
```
Comment (Prithpal-Sooriya) [inline ui/hooks/useMetametrics.ts:40]: perfect decoupling!

**Review by Prithpal-Sooriya — APPROVED** [review]
LGTM, 1 comment to double check:

https://github.com/MetaMask/metamask-extension/pull/29070/files#r1886549170

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -1678,6 +1678,39 @@ export default class MetamaskController extends EventEmitter {
...
+          currState;
+
+        const metaMetricsWasDisabled =
+          prevParticipateInMetaMetrics && !currParticipateInMetaMetrics;
+        const metaMetricsWasEnabled =
+          !prevParticipateInMetaMetrics && currParticipateInMetaMetrics;
+
+        if (!metaMetricsWasDisabled && !metaMetricsWasEnabled) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (Prithpal-Sooriya) [reply app/scripts/metamask-controller.js:1696]: Aha fair enough. Yep we can keep as is.

**Review by jonybur — APPROVED** [review]

**Pull request merged**